File: engine-kernel/src/main/java/com/sos/scheduler/engine/eventbus/AbstractEventBus.java
package com.sos.scheduler.engine.eventbus;

import com.google.common.collect.HashMultimap;
import com.google.common.collect.ImmutableCollection;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import com.google.common.collect.Multimap;
import com.sos.scheduler.engine.data.event.Event;
import com.sos.scheduler.engine.data.event.KeyedEvent;
import com.sos.scheduler.engine.eventbus.annotated.MethodEventSubscriptionFactory;
import java.util.Map;
import java.util.Set;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import static com.google.common.collect.Maps.newHashMap;
import static com.google.common.collect.Sets.newHashSet;

public abstract class AbstractEventBus implements EventBus {
    private static final Logger logger = LoggerFactory.getLogger(AbstractEventBus.class);

    private final Multimap<Class<? extends KeyedEvent<Event>>, EventSubscription> subscribers = HashMultimap.create();
    private final AnnotatedHandlerFinder handlerFinder;
    private final Multimap<EventHandlerAnnotated,EventSubscription> annotatedEventSubscriberMap = HashMultimap.create();
    private final Map<Class<? extends KeyedEvent<Event>>, ImmutableSet<Class<? extends KeyedEvent<Event>>>> cachedSuperEventClasses = newHashMap();

    protected AbstractEventBus(MethodEventSubscriptionFactory factory) {
        this.handlerFinder = new AnnotatedHandlerFinder(factory);
    }

    @Override public final void registerAnnotated(EventHandlerAnnotated o) {
        unregisterAnnotated(o);
        Iterable<EventSubscription> subscribers = handlerFinder.handlers(o);
        annotatedEventSubscriberMap.putAll(o, subscribers);
        registerAll(subscribers);
    }

    @Override public final void unregisterAnnotated(EventHandlerAnnotated o) {
        unregisterAll(annotatedEventSubscriberMap.get(o));
        annotatedEventSubscriberMap.removeAll(o);
    }

    private void registerAll(Iterable<EventSubscription> subscribers) {
        for (EventSubscription s: subscribers)
            register(s);
    }

    private void unregisterAll(Iterable<EventSubscription> subscribers) {
        for (EventSubscription s: subscribers)
            unregister(s);
    }

    public final synchronized void register(EventSubscription s) {
        subscribers.put(s.eventClass(), s);
    }

    public final void unregister(EventSubscription s) {
        boolean ok;
        synchronized(this) {
            ok = subscribers.remove(s.eventClass(), s);
        }
        if (!ok) logger.debug("unregister unknown '" + s + "'");
    }

    public final boolean isSubscribed() {
        return !subscribers.isEmpty();
    }

    protected final synchronized ImmutableCollection<Call> calls(KeyedEvent<Event> e) {
//        Event realEvent = e.event() instanceof EventSourceEvent<?>
//            ? ((EventSourceEvent<Event>)e.event()).event()
//            : (Event)e.event();
//        Class<? extends Event> realEventClass = realEvent.getClass();
        Class<? extends KeyedEvent<Event>> realEventClass = (Class<? extends KeyedEvent<Event>>)e.getClass();
        ImmutableList.Builder<Call> result = null;
        for (Class<? extends KeyedEvent<Event>> c: allSuperEventClasses(realEventClass)) {
            for (EventSubscription s: subscribers.get(c)) {
                // matches() sollte nicht in einer Schleife aufgerufen werden. Das kann langsam werden. Stattdessen sollte die EventSource-Klasse direkt in einer Map abgelegt werden.
                if (matches(s, e)) {
                    if (result == null) result = ImmutableList.builder();
                    result.add(new Call(e, s));
                }
            }
        }
        if (result == null) return ImmutableList.of();
        else return result.build();
    }

    private static boolean matches(EventSubscription s, KeyedEvent<Event> e) {
        return true;  // KeyEvent is the only event class
//        return !(s instanceof EventSourceMethodEventSubscription) ||
//                e instanceof EventSourceEvent &&
//                        ((EventSourceMethodEventSubscription)s).eventSourceMatches((EventSourceEvent<?>)e);
    }

    public final boolean publishNow(KeyedEvent<Event> e) {
        boolean published = false;
        for (Call call: calls(e)) {
            dispatchCall(call);
            published = true;
        }
        return published;
    }

    protected final boolean dispatchCall(Call call) {
        try {
            call.apply();
            return true;
        } catch (Throwable t) {   // Der C++-Code soll wirklich keine Exception bekommen.
            if (t instanceof Error && t.getClass().getSimpleName().equals("TestError"))
                logger.debug("{}", call, t);
            else
                logger.error("{}", call, t);
            //Löst ein rekursives Event aus: log().error(s+": "+x);
            if (EventHandlerFailedEvent.class.isAssignableFrom(call.getEvent().event().getClass())) {
                return false;
            } else
                return publishNow(KeyedEvent.of(new EventHandlerFailedEvent(call, t)));
        }
    }

    private ImmutableSet<Class<? extends KeyedEvent<Event>>> allSuperEventClasses(Class<? extends KeyedEvent<Event>> eventClass) {
        ImmutableSet<Class<? extends KeyedEvent<Event>>> result = cachedSuperEventClasses.get(eventClass);
        if (result == null) {
            Set<Class<? extends KeyedEvent<Event>>> classes = newHashSet();
            collectAllSuperEventClasses(classes, eventClass);
            result = ImmutableSet.copyOf(classes);
            cachedSuperEventClasses.put(eventClass, result);
        }
        return result;
    }

    private static void collectAllSuperEventClasses(Set<Class<? extends KeyedEvent<Event>>> result, Class<?> clas) {
        if (clas != null && KeyedEvent.class.isAssignableFrom(clas)) {
            @SuppressWarnings("unchecked") Class<? extends KeyedEvent<Event>> eventClass = (Class<? extends KeyedEvent<Event>>)clas;
            if (!result.contains(eventClass)) {
                result.add(eventClass);
                collectAllSuperEventClasses(result, clas.getSuperclass());
                for (Class<?> c: clas.getInterfaces())
                    collectAllSuperEventClasses(result, c);
            }
        }
    }
}

File: engine-kernel/src/main/scala/com/sos/scheduler/engine/eventbus/AbstractEventBus.scala
package com.sos.scheduler.engine.eventbus

import com.google.common.collect.{HashMultimap, Multimap}
import com.sos.jobscheduler.common.scalautil.Logger
import com.sos.jobscheduler.data.event.{AnyKeyedEvent, Event, KeyedEvent}
import com.sos.scheduler.engine.eventbus.AbstractEventBus.{logger, matches}
import com.sos.scheduler.engine.eventbus.annotated.MethodEventSubscriptionFactory
import scala.collection.JavaConversions._
import scala.collection.immutable.VectorBuilder

trait AbstractEventBus extends EventBus {
  protected def factory: MethodEventSubscriptionFactory

  final private val subscriptionRegister: Multimap[Class[_ <: Event], EventSubscription] = HashMultimap.create()
  final private val handlerFinder = new AnnotatedHandlerFinder(factory)
  final private val annotatedEventSubscriberMap: Multimap[EventHandlerAnnotated, EventSubscription] = HashMultimap.create()
  private final val superEventClasses = new CachedSuperEventClasses

  final def registerAnnotated(o: EventHandlerAnnotated): Unit = {
    unregisterAnnotated(o)
    val subscribers = handlerFinder.handlers(o)
    annotatedEventSubscriberMap.putAll(o, subscribers)
    registerAll(subscribers)
  }

  final def unregisterAnnotated(o: EventHandlerAnnotated): Unit = {
    unregisterAll(annotatedEventSubscriberMap.get(o))
    annotatedEventSubscriberMap.removeAll(o)
  }

  private def registerAll(subscribers: Iterable[EventSubscription]): Unit = {
    for (s ← subscribers) subscribe(s)
  }

  private def unregisterAll(subscribers: Iterable[EventSubscription]): Unit = {
    for (s ← subscribers) unsubscribe(s)
  }

  final def subscribe(s: EventSubscription): Unit = {
    synchronized {
      subscriptionRegister.put(s.eventClass, s)
    }
  }

  final def unsubscribe(s: EventSubscription): Unit = {
    val ok = synchronized {
      subscriptionRegister.remove(s.eventClass, s)
    }
    if (!ok) logger.debug(s"unsubscribe unknown '$s'")
  }

  final def isSubscribed: Boolean = !synchronized { subscriptionRegister.isEmpty }

  final protected def calls(e: AnyKeyedEvent): Vector[Call] = {
    var result: VectorBuilder[Call] = null
    val classes = superEventClasses(e.event.getClass)
    synchronized {
      for (c ← classes;
           s ← subscriptionRegister.get(c) if matches(s, e)) {
        if (result == null) result = new VectorBuilder[Call]()
        result += Call(e, s)
      }
    }
    if (result != null) result.result else Vector()
  }

  final protected def dispatchCall(call: Call): Boolean =
    try {
      call.apply()
      true
    }
    catch {
      case t: Throwable ⇒  // Der C++-Code soll wirklich keine Exception bekommen.
        if (t.isInstanceOf[Error] && t.getClass.getSimpleName == "TestError")
          logger.debug(s"$call", t)
        else
          logger.error(s"$call", t)
        call.keyedEvent.event match {
          case _: EventHandlerFailedEvent ⇒ false
          case _ ⇒ publishNow(KeyedEvent(new EventHandlerFailedEvent(call, t)))
        }
    }

  private def publishNow(e: AnyKeyedEvent): Boolean = {
    var published = false
    for (call ← calls(e)) {
      dispatchCall(call)
      published = true
    }
    published
  }

}

object AbstractEventBus {
  private val logger = Logger(getClass)

  private def matches(s: EventSubscription, e: AnyKeyedEvent): Boolean = {
    true // KeyEvent is the only event class
    //        return !(s instanceof EventSourceMethodEventSubscription) ||
    //                e instanceof EventSourceEvent &&
    //                        ((EventSourceMethodEventSubscription)s).eventSourceMatches((EventSourceEvent<?>)e);
  }
}
